README.md
## Bearcat-dao -- a SQL mapping dao framework  

## Overview
[bearcat-dao](https://github.com/bearcatjs/bearcat-dao) is a DAO (domain access objects) framework for [node.js](http://nodejs.org/). It implements SQL mapping as its main concept compared to O/R mapping, therefore SQL is still the main concern using with bearcat-dao, and bearcat-dao will map the datebase resultset into [bearcat model](http://bearcatjs.org/guide/model.html) object.  

## SQL mapping vs O/R mapping
Structured Query Language (SQL) has been around for a long time, relational database and SQL have been claimed to have stood the test of time. Moreover, we have experiences whereby the database and even the SQL itself have outlived the application source code, and even mulitiple versions of the source code.  
SQL mapping is on the idea that there is value in relational database and SQL, developers write SQL and maps data resultsets into objects. Therefore, it is easy for enterprise application to optimize, reuse SQL, maintain.  
In another way, O/R mapping enables developers to write mapping object to database table, ORM framework then generates the specific SQL to execute on the database. So, as we can see, developers have to take great knowledge of the ORM framework in order to use the database well, especially when optimization is needed.  

## Model
model definition is using [bearcat model](http://bearcatjs.org/guide/model.html)  
therefore it is easy to be mapped into table and setup constraint, relation  

for example, if we have a test table with single primary id  

```
create table test(
    id bigint(20) NOT NULL COMMENT 'id',	
    
    PRIMARY KEY (id)
)ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

the we can define the following model  

```
var TestModel = function() {
    this.$mid = "testModel";
    this.$table = "test";
    this.id = "$primary;type:Number";
}
  
module.exports = TestModel;
```

in the ***TestModel***, we use ***$table*** attribute to setup the mapping table name, in ***id*** attribute we use ***primary*** to mark it as a primary key, then we add with a type constraint    

## Relation
Tables in relational database can have relations, there are one-to-one relation, one-to-many relation, many-to-many relation  

### One-to-one relation
One-to-one relation means in two models, one model has the reference of the other model  

for example, if we have a ***test1*** table with primary id and reference id of the ***test2*** table  

```
create table test1(
    id bigint(20) NOT NULL COMMENT 'id',	
    rid bigint(20) NOT NULL COMMENT 'reference to test2 id',	
      
    PRIMARY KEY (id)
)ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

```
create table test2(
    id bigint(20) NOT NULL COMMENT 'id',	
    
    PRIMARY KEY (id)
)ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

then we can define the following two models  

```
var Test1Model = function() {
    this.$mid = "test1Model";
    this.$table = "test1";
    this.id = "$primary;type:Number";
    this.test2 = "$type:Object;ref:test2Model"
}
  
module.exports = Test1Model;
```

```
var Test2Model = function() {
    this.$mid = "test2Model";
    this.$table = "test2";
    this.id = "$primary;type:Number";
}
  
module.exports = Test2Model;
```

as we can see, in ***Test1Model.test2*** attribute we use ***ref:test2Model*** to set the reference to ***test2Model***  

### One-to-many relation
One-to-many relation means one model refer to the array of other model. In the real world, for example, we can have a blog, and a blog have many commnets, so blog and comment are one-to-many relation.  

```
var Test1Model = function() {
    this.$mid = "test1Model";
    this.$table = "test1";
    this.id = "$primary;type:Number";
    this.test2 = "$type:Array;ref:test2Model"
}
  
module.exports = Test1Model;
```

therefore, in the above example, we simply modify the ***test2*** attribute type to ***Array***, it becomes a one-to-many relation  

### Many-to-many relation
many-to-many relation can be spilted into two one-many relation through middle table  

## SQL template
When writing complex sql, it is not quite well writing as a String, the better way is using SQL template.  

write SQL tempalte is easy  

for example, we can define SQL template with id ***testResultSql***

```
sqlstart testResultSql
select * from test 
sqlend
```

then we can use this sql in dao, like this  

```
domainDaoSupport.getList("$testResultSql", null, "testModel", function(err, results) {
     // results is testModel type array
});
```

in domainDaoSupport.getList api, the first argument can be SQL tempalte id, the second argument is the SQL arguments, the third argument is the SQL result mapping model id, then in the callback function, we can get the results which are already mapped with testModel array  

Moreover, SQL template can include other SQL template

for example  
```
sqlstart testResultSql
select * from ${testResultTable} 
sqlend

sqlstart testResultTable
test
sqlend
```

then testResultSql template is equal to the above 

## ResultSet mapping
ResultSet is an array of field/value objects, therefore the process of mapping resultSet is like filling objects with specific key/value pairs, to make the key match with the resultSet, we can use ***prefix*** in [model magic attribute value](http://bearcatjs.org/guide/model.html#model_magic_attribute_value) or use ***prefix*** in [model attribute](http://bearcatjs.org/guide/magic-javaScript-objects-in-details.html) to mark all attributes in this model will be prefixed  

for example, if you query for a resultSet like this  

```
[{
	"id": 1,
	"title": "blog_title",
	"content": "blog_content",
	"create_at": 1234567,
	"update_at": 1234567
}]
```

then mapping model can be like this  

```
var BlogModel = function() {
    this.$mid = "blogModel";
    this.$table = "ba_blog";
    this.id = "$primary;type:Number";
    this.aid = "$type:Number";
    this.title = "$type:String";
    this.content = "$type:String";
    this.create_at = "$type:Number";
    this.update_at = "$type:Number";
}
  
module.exports = BlogModel;
```

if your resultSet is prefixed with ***blog_*** like this  
```
[{
	"blog_id": 1,
	"blog_title": "blog_title",
	"blog_content": "blog_content",
	"blog_create_at": 1234567,
	"blog_update_at": 1234567
}]
```

then mapping model will be like this  

```
var BlogModel = function() {
    this.$mid = "blogModel";
    this.$table = "ba_blog";
    this.$prefix = "blog_";
    this.id = "$primary;type:Number";
    this.aid = "$type:Number";
    this.title = "$type:String";
    this.content = "$type:String";
    this.create_at = "$type:Number";
    this.update_at = "$type:Number";
}
  
module.exports = BlogModel;
```

just add ***this.$prefix*** model attribute  

## DAO
DAO is short for domain access object, we can use DAO objects to manage database

bearcat-dao provides ***domainDaoSupport*** wrapping basic sql and cache operations  
add it with properties dependency injection, and init it by invoking ***initConfig*** method  
then you can use domainDaoSupport convenient methods to wrap your own daos 

simpleDao.js
```
var SimpleDao = function() {
    this.$id = "simpleDao";
    this.$init = "init";
    this.$domainDaoSupport = null;
}
  
SimpleDao.prototype.init = function() {
    // init with SimpleModel id to set up model mapping
    this.domainDaoSupport.initConfig("simpleModel");
}
  
// query list all
// callback return mapped SimpleModel array results
SimpleDao.prototype.getList = function(cb) {
    var sql = ' 1 = 1';
    this.$domainDaoSupport.getListByWhere(sql, null, null, cb);
}
  
module.exports = SimpleDao;
```

api reference for [domainDaoSupport](http://bearcatjs.github.io/bearcat-dao/domainDaoSupport.js.html)

## Configuration
add bearcat-dao to your project  

```
npm install bearcat-dao --save
```  

modify context.json used by your project  
[placeholds](http://bearcatjs.org/guide/consistent-configuration.html) can be nicely used to switch between contexts  

```
"dependencies": {
	"bearcat-dao": "*"
},
"beans": [{
    "id": "mysqlConnectionManager",
    "func": "node_modules.bearcat-dao.lib.connection.sql.mysqlConnectionManager",
    "props": [{
        "name": "port",
        "value": "${mysql.port}"
    }, {
        "name": "host",
        "value": "${mysql.host}"
    }, {
        "name": "user",
        "value": "${mysql.user}"
    }, {
        "name": "password",
        "value": "${mysql.password}"
    }, {
        "name": "database",
        "value": "${mysql.database}"
    }]
}, {
    "id": "redisConnectionManager",
    "func": "node_modules.bearcat-dao.lib.connection.cache.redisConnectionManager",
    "props": [{
        "name": "port",
        "value": "${redis.port}"
    }, {
        "name": "host",
        "value": "${redis.host}"
    }]
}]
```

if you do not use redis, you can remove ***redisConnectionManager*** definition  

## Transaction

bearcat-dao provides transaction support based on [bearcat AOP](http://bearcatjs.org/guide/aop.html). The aspect is [transactionAspect](https://github.com/bearcatjs/bearcat-dao/blob/master/lib/aspect/transactionAspect.js) which provides around advice, when target transaction method calls cb function with ***err***, rollback will be emited, otherwise it will commit the operations.  

The pointcut defined is:  
```
"pointcut": "around:.*?Transaction$"
```  

Therefore, any POJO method match this pointcut can a transcation method  
Since transaction must be within the same connection, in Bearcat-dao it is ***transactionStatus***, daos under the transaction method must hold the same transactionStatus  
```
SimpleService.prototype.testMethodTransaction = function(cb, txStatus) {
    var self = this;
  
    this.simpleDao1.transaction(txStatus).addPerson(['aaa'], function(err, results) {
        if (err) {
             return cb(err); // if err occur, rollback will be emited
        }
  
        self.simpleDao2.transaction(txStatus).getList([1, 2], function(err, results) {
            if (err) { 
                 return cb(err); // if err occur, rollback will be emited
            }

            cb(null, results); // commit the operations
        });
    });
}
```

## Enable Debug Mode
run with BEARCAT_DEBUG flag true  
```
BEARCAT_DEBUG=true node xxx.js
```

## Examples
- [bearcat-todo](https://github.com/bearcatnode/todo) 
- [bearcat-dao example](https://github.com/bearcatjs/bearcat-examples/tree/master/bearcat-dao-example)

## License

(The MIT License)

Copyright (c) fantasyni and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

history.md
1.0.2 / 2018-03-06
==================
* 修复SQL语句对 SQL/END & sql/end 的支持
* 修复在SQL语句中字段名出现`end`, 例如friend/gender等词中的`end`也会按照语句结束处理的问题

0.2.12 / 2016-1-28
==================
* update mysqlTemplate add directQuery, directAdd interface to be compatible with ddbTemplate

0.2.11 / 2015-10-14
==================
* make domainDaoSupport support $ prefix sql

0.2.10 / 2015-5-24
==================
* redisClusterTemplate add multi command same key transaction

0.2.9 / 2015-5-24
==================
* redisClusterTemplate add multi command support

0.2.5 / 2015-4-28
==================
* fix modelGet get value bug

0.2.4 / 2015-4-24
==================
* fix directPlan get connection bug

0.2.3 / 2015-4-23
==================
* add dependency bearcat-sql

0.2.2 / 2015-4-23
==================
* add mysql and redis cluster client sharding support based on bearcat-ha 

0.2.1 / 2015-3-1
==================
* fix some bugs, make improvements

0.2.0 / 2015-3-1
==================
* release 0.2.0, see [bearcat-dao](http://bearcatjs.org/topic/dao.html)

0.1.x ~ 0.2.0 
================
dao, sql mapping, transaction

test/core/debug.js
var Bearcat = require('bearcatjs');
var personDomain = require('../mock/domain/person');
var joinPersonDomain = require('../mock/domain/joinPerson');
var domainFactory = require('../../lib/util/domainFactory');
var simplepath = require.resolve('../../test-context.json');
var path = require('path');
var paths = [simplepath];
var bearcatDao = require('../../lib/bearcat-dao');
var bearcat = Bearcat.createApp(paths);
var tableName = "bearcat_dao_test";
var sqlPath = require.resolve('../mock/schema.sql');
var sqlDirPath = path.dirname(sqlPath);

bearcatDao.loadSQL([sqlDirPath]);

process.env.LOGGER_LINE = true;
process.env.BEARCAT_DEBUG = true;
bearcat.start(function() {
	var domainDaoSupport = bearcat.getBean('domainDaoSupport');
	domainDaoSupport.initConfig("person");

	var id = 1;
	// domainDaoSupport.getList("$blogResultSql", id, "blogResult", function(err, results) {
	// 	err = err || true;
	// 	console.log(err.stack);

	// 	// console.log(results);
	// 	var blogResult = results;
	// 	blogResult.run();
	// 	// console.log(blogResult);
	// })
	var domainDaoSupport = bearcat.getBean('domainDaoSupport');
	domainDaoSupport.initConfig(personDomain);
	var person1 = domainFactory.getDomain(personDomain);
	person1.setName('yyy');
	person1.setNum(100);
	person1.setCreateAt(Date.now());

	var list = [];
	list.push(person1);

	var person2 = domainFactory.getDomain(personDomain);
	person2.setName('bbb');
	person2.setNum(200);
	person2.setCreateAt(Date.now());

	list.push(person2);

	var person3 = domainFactory.getDomain(personDomain);
	person3.setId(Date.now() / 1000);
	person3.setName('ccc');
	person3.setNum(300);
	person3.setCreateAt(Date.now());

	list.push(person3);

	domainDaoSupport.batchAdd(list, function(err, results) {
		console.log(err);
	});
});

test/template/sql/DDBTemplate.js
var simplepath = require.resolve('../../../test-ddb-context.json');
var bearcatDao = require('../../../lib/bearcat-dao');
var bearcat = require('bearcatjs');
var path = require('path');
var paths = [simplepath];
bearcat.createApp(paths);

process.env.LOGGER_LINE = true;
bearcat.start(function() {
	// directQuery();
	// for (var i = 0; i < 10; i++) {
	// query();
	// }
	// add();
	update();
});

function query() {
	var sqlTemplate = bearcat.getBean('sqlTemplate');
	// var sql = "select blog.id blog_id, title, content, name from blog, author where blog.aid = author.id";
	// var sql = "select id as idd, title from blog group by idd";
	// var sql = "select title from blog where id > 1 group by title, create_at";
	// var sql = "select title from blog where id > 1 order by title asc, create_at";
	var sql = "select id, title from blog where id = 1 or id = 5";
	// var sql = "select id, title from blog where id between 1 and 5";
	// var sql = "select id, title from blog where id between (1 and 5)";
	// var sql = "select id, title from blog where title is null";
	// var sql = "select id, title from blog where id not in (1, 2)";
	// var sql = "select id, title from blog where id > 1";
	// var sql = "select 1";
	sqlTemplate.executeQuery(sql, [], function(err, results) {
		if (err) {
			console.log(err);
		}

		// console.log(results);
	});
}

function directQuery() {
	var sqlTemplate = bearcat.getBean('sqlTemplate');
	// var sql = "select blog.id blog_id, title, content, name from blog, author where blog.aid = author.id";
	// var sql = "select id as idd, title from blog group by idd";
	// var sql = "select title from blog where id > 1 group by title, create_at";
	// var sql = "select title from blog where id > 1 order by title asc, create_at";
	// var sql = "select id, title from blog where id = 1 or id = 5";
	// var sql = "select id, title from blog where id between 1 and 5";
	// var sql = "select id, title from blog where id between (1 and 5)";
	// var sql = "select id, title from blog where title is null";
	// var sql = "select id, title from blog where id not in (1, 2)";
	// var sql = "select id, title from blog where id > 1";
	// var sql = "select 1";
	var sql = "insert into blog (id, aid, title, content, create_at, update_at) values (10, 10, 'test_title_10', 'test_content_10', 1424659231463, 1424659231463)";
	// var sql = "insert into author (id, name ,create_at, update_at) values (10, 'test_author_10', 123456, 123456)";

	var options = {
		destDB: '*',
		role: 'master'
	}

	options['destDB'] = bearcatDao.calDestDBs('blog', 'id', [10]);

	sqlTemplate.directQuery(sql, [], options, function(err, results) {
		if (err) {
			console.log(err.stack);
		}

		var sql = "select * from blog where id = 10";
		console.log(results);

		sqlTemplate.directQuery(sql, [], options, function(err, results) {
			console.log(results);
		});
	});
}

function add() {
	var sqlTemplate = bearcat.getBean('sqlTemplate');

	var options = {
		destDB: '*',
		role: 'master',
		table: 'author'
	}

	var object = null;
	options['destDB'] = bearcatDao.calDestDBs('blog', 'id', [11]);

	var author = {
		id: 11,
		name: 'test_author_11',
		create_at: 123456,
		update_at: 123456
	}

	sqlTemplate.directAdd(author, options, function(err, results) {
		if (err) {
			console.log(err.stack);
		}

		console.log(results);
	});
}

function update() {
	var sqlTemplate = bearcat.getBean('sqlTemplate');

	var options = {
		destDB: '*',
		role: 'master',
		table: 'author'
	}

	var object = null;
	options['destDB'] = bearcatDao.calDestDBs('blog', 'id', [11]);

	var author = {
		id: 11,
		name: 'test_author_11_update',
		create_at: 123456,
		update_at: 123456
	}

	sqlTemplate.directUpdateById(author, options, function(err, results) {
		if (err) {
			console.log(err.stack);
		}

		console.log(results);
	});
}

lib/util/utils.js
/*!
 * .______    _______     ___      .______       ______     ___   .__________.
 * (   _  )  (   ____)   /   \     (   _  )     (      )   /   \  (          )
 * |  |_)  ) |  |__     /  ^  \    |  |_)  )   |  ,----'  /  ^  \ `---|  |---`
 * |   _  <  |   __)   /  /_\  \   |      )    |  |      /  /_\  \    |  |
 * |  |_)  ) |  |____ /  _____  \  |  |)  ----.|  `----./  _____  \   |  |
 * (______)  (_______/__/     \__\ ( _| `.____) (______)__/     \__\  |__|
 *
 * Bearcat-dao Utils
 * Copyright(c) 2015 fantasyni <fantasyni@163.com>
 * MIT Licensed
 */

var crc = require('redis-crc');
var Utils = {};
var bearcat;

/**
 * Utils check type
 *
 * @param  {String}   type
 * @return {Function} high order function
 * @api public
 */
Utils.isType = function(type) {
	return function(obj) {
		return {}.toString.call(obj) == "[object " + type + "]";
	}
}

/**
 * Utils check array
 *
 * @param  {Array}   array
 * @return {Boolean} true|false
 * @api public
 */
Utils.checkArray = Array.isArray || Utils.isType("Array");

/**
 * Utils check number
 *
 * @param  {Number}  number
 * @return {Boolean} true|false
 * @api public
 */
Utils.checkNumber = Utils.isType("Number");

/**
 * Utils check function
 *
 * @param  {Function}   func function
 * @return {Boolean}    true|false
 * @api public
 */
Utils.checkFunction = Utils.isType("Function");
/**
 * Utils check object
 *
 * @param  {Object}   obj object
 * @return {Boolean}  true|false
 * @api public
 */
Utils.checkObject = Utils.isType("Object");

/**
 * Utils check string
 *
 * @param  {String}   string
 * @return {Boolean}  true|false
 * @api public
 */
Utils.checkString = Utils.isType("String");

/**
 * Utils check boolean
 *
 * @param  {Object}   obj object
 * @return {Boolean}  true|false
 * @api public
 */
Utils.checkBoolean = Utils.isType("Boolean");

/**
 * Utils args to array
 *
 * @param  {Object}  args arguments
 * @return {Array}   array
 * @api public
 */
Utils.to_array = function(args) {
	var len = args.length;
	var arr = new Array(len);

	for (var i = 0; i < len; i++) {
		arr[i] = args[i];
	}

	return arr;
}

/**
 * Utils check is not null
 *
 * @param  {Object}   value
 * @return {Boolean}  true|false
 * @api public
 */
Utils.isNotNull = function(value) {
	if (value !== null && typeof value !== 'undefined')
		return true;
	return false;
}

Utils.getBearcat = function() {
	if (bearcat) {
		return bearcat;
	}

	try {
		bearcat = require('bearcatjs');
	} catch (e) {
		console.error(e);
	}
	return bearcat;
}

/**
 * Utils check model filter error
 *
 * @return {Boolean}  true|false
 * @api public
 */
Utils.checkModelFilterError = function(r) {
	return r !== true && this.isNotNull(r);
}

/**
 * Utils calculate hash value
 *
 * @param  {String}  hash key
 * @param  {Number}  num
 * @return {String}  hash value
 * @api public
 */
Utils.calHashValue = function(key, num) {
	var value = crc.keyHashSlot(key, key.length, num);
	return value;
}

module.exports = Utils;

lib/util/constant.js
/*!
 * .______    _______     ___      .______       ______     ___   .__________.
 * (   _  )  (   ____)   /   \     (   _  )     (      )   /   \  (          )
 * |  |_)  ) |  |__     /  ^  \    |  |_)  )   |  ,----'  /  ^  \ `---|  |---`
 * |   _  <  |   __)   /  /_\  \   |      )    |  |      /  /_\  \    |  |
 * |  |_)  ) |  |____ /  _____  \  |  |)  ----.|  `----./  _____  \   |  |
 * (______)  (_______/__/     \__\ ( _| `.____) (______)__/     \__\  |__|
 *
 * Bearcat-dao Constant
 * Copyright(c) 2014 fantasyni <fantasyni@163.com>
 * MIT Licensed
 */

module.exports = {
	DEFAULT_REDIS_PORT: 6379,
	DEFAULT_REDIS_HOST: '127.0.0.1',
	DEFAULT_MYSQL_PORT: 3306,
	DEFAULT_MYSQL_HOST: '127.0.0.1',
	DEFAULT_MYSQL_USE_POOL: true,
	DEFAULT_MYSQL_CONNECT_CB: function() {},
	DEFAULT_REDIS_CONNECT_CB: function() {},

	TYPE_LONG: 'long',
	TYPE_NUMBER: 'number',

	TRANSACTION_DEFAULT: "",
	TRANSACTION_START: "tstart",

	FILE_TYPE_SQL: 'sql',

	COMMENT_LINE_1: /\/\/.*?\n/g,
	COMMENT_LINE_2: /\-\-.*?\n/g,
	COMMENT_STAR: /\/\*(.|\s)*?\*\//g,

    CONTENT_BLOCK: /\bsqlstart\b\s*\w+(.|\s)*?\bsqlend\b/ig,
    CONTENT_BLOCK_SQL: /\bsqlstart\b\s*(\w+)((.|\s)*?)\bsqlend\b/i,
	CONTENT_SQL_TRIM: /\n\s*/g,
	SQL_REF: /\$\{\w+\}/g,
	SQL_REF_ID: /\$\{(\w+)\}/,

	DOLLAR_PREFIX: '$',

	PLAN_QUERY: 'select',

	SQL_SELECT_PART: {
		COLUMN: 'column',
		SOURCE: 'source',
		JOIN: 'joinmap',
		WHERE: 'where',
		GROUPBY: 'groupby',
		ORDERBY: 'orderby',
		LIMIT: 'limit'
	},

	ROLE_MASTER: 'master',
	ROLE_SLAVE: 'slave',

	LEXTER_TYPE_NUMBER: 2,

	SELECT_CONDITION_EQUAL: "=",
	SELECT_CONDITION_IN: "in",
	SELECT_CONDITION_NOT_IN: "not in",
	SELECT_CONDITION_NOT_NULL: "not null",
	SELECT_CONDITION_BETWEEN: "between",
	SELECT_CONDITION_VALUE_TYPE_KEYWORD: 1,
	SELECT_CONDITION_VALUE_TYPE_STRING: 3,
	SELECT_CONDITION_VALUE_TYPE_OPERATOR: 7,
	CLUSTER_MAP_PATH: "clusterSharding.json",
	MYSQL_CLUSTER_MAP_PATH: "mysqlSharding.json",
	REDIS_CLUSTER_MAP_PATH: "redisSharding.json",
	DEST_DB_ALL: "*",

	ORDER_ASC: 1,
	ORDER_DESC: 2,

	REDIS_RETRY_MAX_DELAY: 30 * 1000,

	ID_FIELD: "id",

	REDIS_CLUSTER_COMMANDS: [
		'del', 'expire', 'ttl',
		'get', 'set', 'exists',
		'llen', 'lrem', 'lpush', 'lrange',
		'sadd', 'scard', 'smembers', 'srem',
		'zadd', 'zcard', 'zrangebyscore',
		'zrem', 'zrange', 'zremrangebyrank',
		'zscore'
	]
}